docker/alias/alias_computation.py

```diff
@@ -0,0 +1,135 @@
+    alias_group.key.delete()
+    return
+  if len(bug_ids) > ALIAS_GROUP_VULN_LIMIT:
+    logging.info('Deleting alias group due to too many bugs: %s', bug_ids)
+    alias_group.key.delete()
+    return
+
+  if bug_ids == alias_group.bug_ids:
+    return
+
+  alias_group.bug_ids = bug_ids
+  alias_group.last_modified = datetime.datetime.utcnow()
+  alias_group.put()
+
+
+def _create_alias_group(bug_ids):
+  """Creates a new alias group in the datastore."""
+  if len(bug_ids) <= 1:
+    logging.info('Skipping alias group creation due to too few bugs: %s',
+                 bug_ids)
+    return
+  if len(bug_ids) > ALIAS_GROUP_VULN_LIMIT:
+    logging.info('Skipping alias group creation due to too many bugs: %s',
+                 bug_ids)
+    return
+
+  new_group = osv.AliasGroup(bug_ids=bug_ids)
+  new_group.last_modified = datetime.datetime.utcnow()
+  new_group.put()
+
+
+def _compute_aliases(bug_id, visited):
+  """Computes all aliases for the given bug ID.
+  The returned list contains the bug ID itself, all the IDs from the bug's
+  raw aliases, all the IDs of bugs that have the current bug as an alias,
+  and repeat for every bug encountered here."""
+  to_visit = {bug_id}
+  bug_ids = []
+  while to_visit:
+    bug_id = to_visit.pop()
+    if bug_id in visited:
+      continue
+    visited.add(bug_id)
+    bug_ids.append(bug_id)
+
+    aliases = bugs_map.get(bug_id, [])
+    bugs_of_alias = aliases_map.get(bug_id, [])
+    to_visit.update(set(bugs_of_alias + aliases) - visited)
+
+  return sorted(bug_ids)
+
+
+def main():
+  """Updates all alias groups in the datastore by re-computing existing
+  AliasGroups and creating new AliasGroups for un-computed bugs."""
+
+  # Query for all bugs that have aliases.
+  bugs = osv.Bug.query(osv.Bug.aliases > '')
+  all_alias_group = osv.AliasGroup.query()
+  allow_list_query = osv.AliasAllowListEntry.query()
+  deny_list_query = osv.AliasDenyListEntry.query()
+  allow_list = {allow_entry.bug_id for allow_entry in allow_list_query}
+  deny_list = {deny_entry.bug_id for deny_entry in deny_list_query}
+
+  # For each bug, add its aliases to the maps and ignore invalid bugs.
+  for bug in bugs:
+    if bug.db_id not in allow_list and (bug.db_id in deny_list or
+                                        len(bug.aliases) > VULN_ALIASES_LIMIT):
+      logging.info('%s has too many listed aliases, skipping computation.',
+                   bug.db_id)
+      continue
+    if bug.withdrawn:
+      continue
+    bugs_map[bug.db_id] = bug.aliases
+    for alias in bug.aliases:
+      aliases_map.setdefault(alias, []).append(bug.db_id)
+
+  visited = set()
+
+  # For each alias group, re-compute the bug IDs in the group and update the
+  # group with the computed bug IDs.
+  for alias_group in all_alias_group:
+    bug_id = alias_group.bug_ids[0]  # AliasGroups contain more than one bug.
+    if bug_id in visited:
+      alias_group.key.delete()
```

Comment:
We talked about this offline, but also add a comment to explain why we delete the AliasGroup if we've already visited the bug (i.e. because this means two AliasGroups have been combined)